### PR TITLE
fix(Shard): Add handler & event for parent process death

### PR DESCRIPTION
**Pull request by uhKevinMC**
**Please describe the changes this PR makes and why it should be merged:**

This PR adds an event named `parentDeath` and handler for child processes/workers to terminate upon IPC channel disconnect, aka parent process disconnect or death. This also adds a `timeout` option to the ShardingManager, which sets an interval on each shard to check if the parentPort has been nullified, aka parent thread hangup. This PR was made from issue #4151 

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [ ] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by uhKevinMC** [comment]
It seems like worker threads do not have an issue with zombie processes, so I'll open up this PR as its basically done.

**Comment by uhKevinMC** [comment]
seems like the issue does occur on workers, but at a small chance and if it was terminated manually by OS. Typings I will need help with.

File: src/sharding/Shard.js [hunk 1]
```diff
@@ -117,11 +125,20 @@ class Shard extends EventEmitter {
           execArgv: this.execArgv,
         })
         .on('message', this._handleMessage.bind(this))
-        .on('exit', this._exitListener);
+        .on('exit', this._exitListener)
+        .on('disconnect', this._handleDisconnect.bind({ parentProcessDisconnect: true }));
```
Comment (ItsPapaia) [inline src/sharding/Shard.js:129]:
```suggestion
        .on('disconnect', this._handleDisconnect.bind(this, { parentProcessDisconnect: true }));
```

Comment (uhKevinMC) [reply src/sharding/Shard.js:129]: slight oversight on my part, thanks for the suggestion

**Comment by MrAugu** [comment]
I havent tested the new code after review but it didnt fix the problem, il update it and are the.

**Comment by MrAugu** [comment]
I'm testing the code.

File: src/sharding/Shard.js [hunk 2]
```diff
@@ -117,11 +131,20 @@ class Shard extends EventEmitter {
...
+        .on('disconnect', this._handleDisconnect.bind(this, { parentProcessDisconnect: true }));
     } else if (this.manager.mode === 'worker') {
       this.worker = new Worker(path.resolve(this.manager.file), { workerData: this.env })
         .on('message', this._handleMessage.bind(this))
         .on('exit', this._exitListener);
+
+      const parentPort = require('worker_threads').parentPort;
+      this.parentPort = parentPort;
+      setInterval(() => {
+        if (!parentPort) {
```
Comment (NotSugden) [inline src/sharding/Shard.js:144]: wouldn't this always be truthy? the variable is never re-assigned as its a const
and should the interval be cleared if this condition is true?

File: src/sharding/Shard.js [hunk 3]
```diff
@@ -77,6 +77,12 @@ class Shard extends EventEmitter {
      */
     this.worker = null;
 
+    /**
+     * Parent port of the worker threads (if {@link ShardingManager#mode} is `worker`)
+     * @type (?number)
```
Comment (NotSugden) [inline src/sharding/Shard.js:82]:
```suggestion
     * @type {?number}
```
wrong brackets

File: src/sharding/Shard.js [hunk 4]
```diff
@@ -97,6 +103,14 @@ class Shard extends EventEmitter {
...
      */
     this._exitListener = this._handleExit.bind(this, undefined);
+
+    /**
+     * Timeout for worker threads (if {@link ShardingManager#mode} is `worker`)
+     * @type {?number}
+     * @private
+     */
+
+    this.timeout = manager.timeout;
```
Comment (NotSugden) [inline src/sharding/Shard.js:113]:
```suggestion
     */
    this.timeout = manager.timeout;
```
nit

File: src/sharding/Shard.js [hunk 5]
```diff
@@ -117,11 +131,23 @@ class Shard extends EventEmitter {
...
+
+      this.parentPort = require('worker_threads').parentPort;
+
+      if (this.parentPort) {
+        this.manager.client.setInterval(() => {
+          this.parentPort = require('worker_threads').parentPort;
+          if (!this.parentPort) {
+            this._handleDisconnect({ parentThreadDisconnect: true });
+          }
+        }, this.timeout);
```
Comment (NotSugden) [inline src/sharding/Shard.js:149]:
```suggestion
        const interval = this.manager.client.setInterval(() => {
          this.parentPort = require('worker_threads').parentPort;
          if (!this.parentPort) {
            this._handleDisconnect({ parentThreadDisconnect: true });
            this.manager.client.clearInterval(interval);
          }
        }, this.timeout);
```
interval should be cleared

**Comment by MrAugu** [comment]
After testing in-production, it still doesn't seem to exactly work.

File: src/sharding/Shard.js (hunk 5 above)
Comment (uhKevinMC) [reply src/sharding/Shard.js:148]: Clearing the interval would be redundant, as the line above will shut down the thread/process regardless of state.

**Comment by uhKevinMC** [comment]
@MrAugu hey, if you can, can you please join the [Discord server](https://discord.gg/bRCvFy9) and send me (KevinMC#0001) the diagnostic information regarding the issue? Such as logs, code, usage information, system information, and any other info (debug, system logs/events, etc...). The more the better.

This is most likely an issue with your operating system or external programs handling the processes, as this is tracking if the parent-related property on each mode has closed/disconnected internally.

I'll review the code once more and do more tests if necessary, but please send over the information as mentioned above.

Thanks,

File: src/sharding/Shard.js (hunk 5 above)
Comment (NotSugden) [reply src/sharding/Shard.js:148]: the interval isnt set on the actual child process is it? it looks as if its being set on the main process?

Comment (uhKevinMC) [reply src/sharding/Shard.js:148]: seems that way, could be why previous attempts at testing the PR went 👎 , I'll see about changing this.

**Comment by kyranet** [comment]
@uhKevinMC did you have any luck getting this to fully work, or do you need help testing the PR?

**Comment by uhKevinMC** [comment]
> @uhKevinMC did you have any luck getting this to fully work, or do you need help testing the PR?

I haven't heard a response from the issue creator so I guess it's solved? It'll be helpful if I had a tester to confirm this.

:)

File: src/sharding/Shard.js [hunk 6]
```diff
@@ -378,6 +410,38 @@ class Shard extends EventEmitter {
...
+   */
+  _handleDisconnect(message) {
+    /**
+     * Emmited upon the shard's parent process exiting.
+     * @event Shard#parentDeath
+     * @param {ChildProcess} process Child process that exited.
+     * This event is internal and should not be emitted manually.
+     * You can listen to this event if you wish.
+     */
+    if (message) {
```
Comment (kyranet) [inline src/sharding/Shard.js:427]: This could short-circuit to reduce indentation, doing `if (!message) return;`.

Comment (kyranet) [reply src/sharding/Shard.js:427]: Additionally, the jsdoc for the event should be moved before this condition.

**Review by kyranet — CHANGES_REQUESTED** [review]

**Comment by uhKevinMC** [comment]
In the interest of y'all not trying to deal with the mess I've created, I have reforked the repo and redid the PR.

**Pull request closed without merging**